File: day09/solver.hs
{-# LANGUAGE QuasiQuotes #-}

import Text.RawString.QQ 
import Data.List
import Utils
import AoC

getInput :: IO String
getInput = readFile "input.txt"

--------------------------------------------------------------------

type Parsed = [[Int]]
type Sol1 = Int
type Sol2 = Int

parse :: String -> Parsed
parse = map (map read . words) . lines

extrapolate :: Int -> [Int] -> Int
extrapolate acc nums =
    if all (==0) nums then
        acc
    else extrapolate (acc + (last nums)) (diff nums)

solve1 :: Parsed -> Sol1
solve1 = sum . map (extrapolate 0)

solve2 :: Parsed -> Sol2
solve2 = solve1 . map reverse


testdata = [r|0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
|]
testresult1 = 114
testresult2 = 2

--------------------------------------------------------------------

test1 = test (solve1 . parse) testresult1 testdata
test2 = test (solve2 . parse) testresult2 testdata

printPart :: (Show a) => Int -> a -> IO ()
printPart part solution = do
    putStr $ "Part " ++ (show part) ++ ": "
    print solution

main' :: CLIOptions -> IO ()
main' (CLIOptions p False) = do
    parsed <- parse <$> getInput
    case p of
        1 -> printPart 1 (solve1 parsed)
        2 -> printPart 2 (solve2 parsed)
        _ -> (do
            printPart 1 (solve1 parsed)
            printPart 2 (solve2 parsed))
main' (CLIOptions p True) = 
    case p of
        1 -> printPart 1 test1
        2 -> printPart 2 test2
        _ -> (do
            printPart 1 test1
            printPart 2 test2)

main = main' =<< clioptions

File: day09/Solver.jl
module Solver
using Test
using AoC
using AoC.Utils
import Base: diff


function parse_input(raw_data)
    map(ints, lines(raw_data|>strip))
end
export parse_input


function solve1(parsed)
    result = 0
    for nums in parsed
        extrapolated = nums[end]
        diffs = diff(nums)
        extrapolated += diffs[end]
        while !all(diffs .== 0)
            diffs = diff(diffs)
            extrapolated += diffs[end]
        end
        result += extrapolated
    end
    result
end
export solve1


function solve2(parsed)
    result = 0
    for nums in parsed
        extrapolated = nums[1]
        diffs = diff(nums)
        # the calculation for `extrapolated` has the form
        # a - (b - (c - (d - e))) which can be simplified to
        # a - b + c - d + e so when iterating over the layers of diffs,
        # the sign of the number to be added must oscillate
        neg = -1
        extrapolated += neg*diffs[1]
        while !all(diffs .== 0)
            diffs = diff(diffs)
            neg *= -1
            extrapolated += neg*diffs[1]
        end
        result += extrapolated
    end
    result
end
export solve2


solution = Solution(parse_input, solve1, solve2)

testinput = """0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
"""
testanswer_1 = 114
testanswer_2 = 2
export testinput, testanswer_1, testanswer_2

test() = AoC.test_solution(solution, testinput, testanswer_1, testanswer_2)
export test

main(part=missing) = AoC.main(solution, part)
export main


end # module Solver
